### PR TITLE
Lower bound improvement based on sharpness of the objective

### PR DESCRIPTION
**Pull request by dhendryc**
I added the bound improvement using sharpness like outline in the paper.

Written simple test on the Normbox (alias Approx Planted Point) example. 

It seems one has to use the **node** dual gap for the sharpness improvement. I left the code for the root dual gap in for now until we have rechecked the proof in the paper.

### CODE REVIEW

File: src/custom_bonobo.jl [hunk 1]
```diff
@@ -75,6 +75,25 @@ function Bonobo.optimize!(
         Bonobo.branch!(tree, node)
         callback(tree, node)
     end
+    # To make sure that we collect the statistics in case the time limit is crashed.
```
Comment (matbesancon) [inline src/custom_bonobo.jl:78]:
```suggestion
    # To make sure that we collect the statistics in case the time limit is reached.
```

**Comment by dhendryc** [comment]
Halved the sharpness constant `M` analogously to the strong convexity case.

File: src/interface.jl [hunk 2]
```diff
@@ -43,6 +43,8 @@ dual_tightening        - whether to use dual tightening techniques (make sure yo
 global_dual_tightening - dual tightening maintained globally valid (when new solutions are found)
 bnb_callback           - an optional callback called at every node of the tree, for example for heuristics
 strong_convexity       - strong convexity of the function, used for tighter dual bound at every node
+sharpness_constant     - the constant M > 0 for (θ, M)-sharpness.
```
Comment (matbesancon) [inline src/interface.jl:46]: we should add the expression we use for sharpness (it can differ from one author to the next)

Comment (dhendryc) [reply src/interface.jl:46]: I added a short version of the definition that we have in the paper.

File: src/callbacks.jl [hunk 3]
```diff
@@ -1,4 +1,11 @@
-# FW callback
+"""
+Frank-Wolfe Callback.
+
+Is called in every Frank-Wolfe iteration.
+Node evaluation can be dymamically stopped here.
```
Comment (matbesancon) [inline src/callbacks.jl:5]:
```suggestion
Node evaluation can be dynamically stopped here.
```

**Review by matbesancon — DISMISSED** [review]

**Review by matbesancon — DISMISSED** [review]

File: src/tightenings.jl [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+            @assert num_fractional == 0 || sharpness_bound >= lower_bound "$(num_fractional) == 0 || $(sharpness_bound) > $(lower_bound)"
+        end
+
+        lower_bound = max(strong_convexity_bound, sharpness_bound)
+    end
+    return lower_bound
+end
+
+"""
+Use strong convexity and/or sharpness to potentially remove one of the children nodes
```
Comment (matbesancon) [inline src/tightenings.jl:232]:
```suggestion
Use strong convexity and/or sharpness to potentially remove one of the children nodes.
If both sharpness and strong convexity parameters are provided, strong convexity is preferred.
```

**Pull request merged**